Port various tasks to .NET Core

dsplaisted opened the pull request:
This adds a variety of tasks to the .NET Core build.  This should be sufficient to compile basic libraries and console applications targeting the .NET Framework with the .NET Core version of MSBuild.

rainersigwald commented on src/Shared/Compat/TypeExtensions.cs:
```diff
@@ -8,5 +12,622 @@ public static bool IsEquivalentTo(this Type type, Type other)
+        }
+
+
+        class InvokeMemberHelper
```
I assume/hope you didn't write all of this just now :smile: Comment source so we can have hope of removing it later or at least keeping it in sync?

rainersigwald commented on src/XMakeTasks/AppConfig/AppConfig.cs:
```diff
@@ -24,27 +24,47 @@ internal sealed class AppConfig
         {
-            XmlTextReader reader = null;
+            
+            XmlReader reader = null;
```
Whitespace

rainersigwald commented on src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs:
```diff
@@ -51,6 +58,7 @@ internal AssemblyInformation(string sourceFile)
             _sourceFile = sourceFile;
 
+#if FEATURE_ASSEMBLY_LOADFROM
             if (NativeMethodsShared.IsWindows)
```
Do we need this check any more, or is it superseded by `FEATURE_ASSEMBLY_LOADFROM`?

dsplaisted replied:
@rainersigwald

> Do we need this check any more, or is it superseded by `FEATURE_ASSEMBLY_LOADFROM`?

I think we still need it for Mono, which supports Assembly.LoadFrom.

rainersigwald commented on src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs:
```diff
@@ -317,8 +329,102 @@ private FrameworkName GetFrameworkName()
+        {
+            // TODO: Nick Guerrera (Nick.Guerrera@microsoft.com) hacked this method for temporary use.
+            // There is a blob decoder feature in progress but it won't ship in time for our milestone.
+            // Replace this method with the blob decoder feature when later it is availale.
```
Sanity check: this isn't already available, right?

dsplaisted replied:
I actually talked with Nick about this.  The functionality is available in an alpha version of System.Reflection.Metadata, but the APIs still need to be reviewed and will probably change, so I preferred to use code that wouldn't break when that happens.

rainersigwald replied:
Totally reasonable.  Can we put a date or version describing when we expect to see the cleaner way available so A Future Maintainer™ doesn't have to do the same research?

rainersigwald commented on src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs:
```diff
@@ -477,6 +622,7 @@ private string[] ImportFiles()
                 return new string[0];
```
Superseded by feature flag?

dsplaisted replied:
Again, not for Mono.

rainersigwald commented on src/XMakeTasks/Microsoft.Build.Tasks.csproj:
```diff
@@ -197,264 +201,279 @@
       <ExcludeFromStyleCop>true</ExcludeFromStyleCop>
     </Compile>
-    <Compile Include="Message.cs">
+    <Compile Include="AppConfig\*.cs">
```
I'm not a fan of wildcards in project files.

dsplaisted replied:
Yes, I just moved it into the section to be compiled for .NET Core.

rainersigwald commented:
Many comments, but overall LGTM.

AndyGerlicher commented:
LGTM as well but I also would like to see the source listed, especially on the TypeExtension class. I assume that was based on something else since it has a large chunk commented out?

dsplaisted commented:
I've addressed the feedback (added comments about where source code was copied from, when there would be an API in System.Reflection.Metadata to decode attribute arguments, and removed an extra blank line).
